src/mlpack/methods/ann/rbm/rbm_impl.hpp

```diff
@@ -0,0 +1,351 @@
+
+// In case it hasn't been included yet.
+#include "rbm.hpp"
+
+#include <mlpack/methods/ann/activation_functions/logistic_function.hpp>
+
+namespace mlpack {
+namespace ann /** Artificial neural networks. */ {
+
+template<
+  typename InitializationRuleType,
+  typename DataType,
+  typename PolicyType
+>
+RBM<InitializationRuleType, DataType, PolicyType>::RBM(
+    arma::Mat<ElemType> predictors,
+    InitializationRuleType initializeRule,
+    const size_t visibleSize,
+    const size_t hiddenSize,
+    const size_t batchSize,
+    const size_t numSteps,
+    const size_t negSteps,
+    const size_t poolSize,
+    const ElemType slabPenalty,
+    const ElemType radius,
+    const bool persistence):
+    predictors(std::move(predictors)),
+    initializeRule(initializeRule),
+    visibleSize(visibleSize),
+    hiddenSize(hiddenSize),
+    batchSize(batchSize),
+    numSteps(numSteps),
+    negSteps(negSteps),
+    poolSize(poolSize),
+    slabPenalty(slabPenalty),
+    radius(2 * radius),
+    persistence(persistence),
+    reset(false)
+{
+  numFunctions = this->predictors.n_cols;
+}
+
+template<
+  typename InitializationRuleType,
+  typename DataType,
+  typename PolicyType
+>
+template<typename Policy>
+typename std::enable_if<std::is_same<Policy, BinaryRBM>::value, void>::type
+RBM<InitializationRuleType, DataType, PolicyType>::Reset()
+{
+  size_t shape = (visibleSize * hiddenSize) + visibleSize + hiddenSize;
+
+  parameter.set_size(shape, 1);
+  positiveGradient.set_size(shape, 1);
+  negativeGradient.set_size(shape, 1);
+  tempNegativeGradient.set_size(shape, 1);
+  negativeSamples.set_size(visibleSize, batchSize);
+
+  weight = DataType(parameter.memptr(), hiddenSize, visibleSize, false, false);
+  hiddenBias = DataType(parameter.memptr() + weight.n_elem,
+      hiddenSize, 1, false, false);
+  visibleBias = DataType(parameter.memptr() + weight.n_elem +
+      hiddenBias.n_elem, visibleSize, 1, false, false);
+
+  parameter.zeros();
+  positiveGradient.zeros();
+  negativeGradient.zeros();
+  tempNegativeGradient.zeros();
+  initializeRule.Initialize(parameter, parameter.n_elem, 1);
+
+  reset = true;
+}
+
+template<
+  typename InitializationRuleType,
+  typename DataType,
+  typename PolicyType
+>
+template<typename OptimizerType>
+void RBM<InitializationRuleType, DataType, PolicyType>::Train(
+    OptimizerType& optimizer)
+{
+  if (!reset)
+  {
+    Reset();
+  }
+
+  optimizer.Optimize(*this, parameter);
+}
+
+template<
+  typename InitializationRuleType,
+  typename DataType,
+  typename PolicyType
+>
+template<typename Policy>
+typename std::enable_if<std::is_same<Policy, BinaryRBM>::value, double>::type
+RBM<InitializationRuleType, DataType, PolicyType>::FreeEnergy(
+    arma::Mat<ElemType>&& input)
+{
+  preActivation = (weight * input);
+  preActivation.each_col() += hiddenBias;
+  preActivation = arma::log(1 + arma::trunc_exp(preActivation));
+  return -(arma::accu(preActivation) + arma::dot(input, arma::repmat(
```

Comment:
Not sure the compiler will inline the expression (``arma::log(1 + arma::trunc_exp(preActivation));``) to avoid another copy.

Reply:
I'll change this to directly return the expression, just to be sure.